File: Java/75 Sort Colors.java
class Solution {
    public void sortColors(int[] nums) {
        int sorted = 0, blues = 0, whites = 0;

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {
                nums[sorted++] = nums[i];
            } else if (nums[i] == 1) {
                whites++;
            } else {
                blues++;
            }
        }

        while (whites > 0) {
            nums[sorted++] = 1;
            whites--;
        }

        while (blues > 0) {
            nums[sorted++] = 2;
            blues--;
        }
    }
}

File: C#/75. Sort Colors.cs
// Problem: https://leetcode.com/problems/sort-colors/

public class Solution {
    // One-pass, constant extra space implementation
    public void SortColors(int[] nums) {
        int i = 0, r = 0, b = nums.Length - 1;

        while (i <= b) {
            if (nums[i] == 0) {
                (nums[i], nums[r]) = (nums[r], nums[i]);
                if (i <= ++r) i++;

            } else if (nums[i] == 1) {
                i++;

            } else if (nums[i] == 2) {
                (nums[i], nums[b]) = (nums[b--], nums[i]);
            }
        }
    }
}
